File: app/src/main/java/com/afterroot/photos/async/DeleteFromDatabaseTask.java
package com.afterroot.photos.async;

import android.content.Context;
import android.database.Cursor;
import android.os.AsyncTask;

import com.afollestad.materialdialogs.MaterialDialog;
import com.afterroot.photos.DatabaseHelper;
import com.afterroot.photos.Helper;
import com.afterroot.photos.adapter.ImagesRecyclerViewAdapter;
import com.afterroot.photos.async.AddToDatabaseTask.TaskCallback;

import java.io.File;
import java.util.List;

public class DeleteFromDatabaseTask extends AsyncTask<List<Integer>, String, String> {
    private Cursor mCursor;
    private Helper mHelper;
    private ImagesRecyclerViewAdapter mAdapter;
    private TaskCallback mRemoveTaskCallback;
    private MaterialDialog mProgressDialog;
    private Context mContext;

    public DeleteFromDatabaseTask(Context context, ImagesRecyclerViewAdapter adapter, TaskCallback removeTaskCallback){
        this.mContext = context;
        mHelper = new Helper(mContext);
        this.mAdapter = adapter;
        this.mRemoveTaskCallback = removeTaskCallback;
    }

    @Override
    protected void onPreExecute() {
        String[] projection = {
                DatabaseHelper.TableColumns.COLUMN_NAME_FILENAME,
                DatabaseHelper.TableColumns.COLUMN_NAME_PATH
        };
        mCursor = mHelper.getCursor(DatabaseHelper.TableColumns.TABLE_NAME_IMAGES, projection);

        mProgressDialog = new MaterialDialog.Builder(mContext)
                .progress(false, mCursor.getCount(), true)
                .title("Removing from Database")
                .content("Loading...")
                .negativeText("Cancel")
                .onNegative((dialog, which) -> cancel(true))
                .cancelable(false)
                .build();
        mProgressDialog.show();

        super.onPreExecute();
    }

    @Override
    protected String doInBackground(List<Integer>[] lists) {
        mCursor.moveToFirst();
        for (int i : lists[0]){
            File file = new File(mAdapter.getPath(i));
            mCursor.moveToNext();
            mHelper.getDatabase().delete(DatabaseHelper.TableColumns.TABLE_NAME_IMAGES,
                    DatabaseHelper.TableColumns.COLUMN_NAME_PATH + "=? AND "+ DatabaseHelper.TableColumns.COLUMN_NAME_TAG + "=?",
                    new String[]{mAdapter.getPath(i), mHelper.getFilterTag()});
            publishProgress(file.getName());
        }
        if (isCancelled()) {
            return "Removing from Database cancelled";
        } else {
            return "Removed from Database";
        }
    }

    @Override
    protected void onProgressUpdate(String... values) {
        mProgressDialog.setContent(values[0]);
        mProgressDialog.setProgress(mCursor.getPosition());
        super.onProgressUpdate(values);
    }

    @Override
    protected void onPostExecute(String s) {
        mCursor.close();
        mHelper.getDatabase().close();
        mProgressDialog.dismiss();
        mRemoveTaskCallback.onTaskFinished();
        mHelper.showToast(s);
        super.onPostExecute(s);
    }
}

File: app/src/main/java/com/afterroot/tagit/async/DeleteFromDatabaseTask.kt
package com.afterroot.tagit.async

import android.content.Context
import android.database.Cursor
import android.os.AsyncTask
import com.afollestad.materialdialogs.MaterialDialog
import com.afterroot.tagit.DatabaseHelper
import com.afterroot.tagit.Helper
import com.afterroot.tagit.TaskCallback
import com.afterroot.tagit.adapter.ImagesRecyclerViewAdapter
import java.io.File

class DeleteFromDatabaseTask(private val mContext: Context, private val mAdapter: ImagesRecyclerViewAdapter, private val mRemoveTaskCallback: TaskCallback) : AsyncTask<List<Int>, String, String>() {
    private lateinit var mCursor: Cursor
    private val mHelper: Helper = Helper(mContext)
    private lateinit var mProgressDialog: MaterialDialog

    override fun onPreExecute() {
        val projection = arrayOf(DatabaseHelper.TableColumns.COLUMN_NAME_FILENAME, DatabaseHelper.TableColumns.COLUMN_NAME_PATH)
        mCursor = mHelper.getCursor(DatabaseHelper.TableColumns.TABLE_NAME_IMAGES, projection)

        mProgressDialog = MaterialDialog.Builder(mContext)
                .progress(false, mCursor.count, true)
                .title("Removing from Database")
                .content("Loading...")
                .negativeText("Cancel")
                .onNegative { _, _ -> cancel(true) }
                .cancelable(false)
                .build()
        mProgressDialog.show()

        super.onPreExecute()
    }

    override fun doInBackground(lists: Array<List<Int>>): String {
        mCursor.moveToFirst()
        for (i in lists[0]) {
            val file = File(mAdapter.getPath(i))
            mCursor.moveToNext()
            mHelper.database!!.delete(DatabaseHelper.TableColumns.TABLE_NAME_IMAGES,
                    DatabaseHelper.TableColumns.COLUMN_NAME_PATH + "=? AND " + DatabaseHelper.TableColumns.COLUMN_NAME_TAG + "=?",
                    arrayOf(mAdapter.getPath(i), mHelper.filterTag))
            publishProgress(file.name)
        }
        return if (isCancelled) {
            "Removing from Database cancelled"
        } else {
            "Removed from Database"
        }
    }

    override fun onProgressUpdate(vararg values: String) {
        mProgressDialog.setContent(values[0])
        mProgressDialog.setProgress(mCursor.position)
        super.onProgressUpdate(*values)
    }

    override fun onPostExecute(s: String) {
        mCursor.close()
        mHelper.database!!.close()
        mProgressDialog.dismiss()
        mRemoveTaskCallback.onTaskFinished()
        mHelper.showToast(s)
        super.onPostExecute(s)
    }
}
